### PR TITLE
ENH: simulation logic

### PR DESCRIPTION
**Pull request by chiahaoliu**
add environment variable logic to enable simulation while keep current pytest alive

in simulation, program help user to spin up local ``db`` so that he can test api and functionalities

### CODE REVIEW

File: xpdan/999-load.py [hunk 1]
```diff
@@ -1,6 +1,91 @@
...
+    def __init__(self, name, read_fields):
+        self.images_per_set = MagicMock()
+        self.images_per_set.get = MagicMock(return_value=5)
+        self.number_of_sets = MagicMock()
+        self.number_of_sets.put = MagicMock(return_value=1)
+        self.number_of_sets.get = MagicMock(return_value=1)
+        self.cam = MagicMock()
+        self.cam.acquire_time = MagicMock()
+        self.cam.acquire_time.put = MagicMock(return_value=0.1)
+        self.cam.acquire_time.get = MagicMock(return_value=0.1)
```
Comment (CJ-Wright) [inline xpdan/999-load.py]: Do we need `MagicMock`? Can we spoof the detector with an actual values?

File: xpdan/tests/conftest.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def build_pymongo_backed_broker(request):
     mds_test_conf = dict(database=db_name, host='localhost',
                          port=27017, timezone='US/Eastern')
     mds = MDS(mds_test_conf, 1,
-              # auth=False
+              auth=False
```
Comment (CJ-Wright) [inline xpdan/tests/conftest.py]: Can we ad a `try, except` here?
```python
try:
    mds = MDS(mds_test_conf, 1, auth=False)
except TypeError:
    mds = MDS(mds_test_conf, 1)

File: xpdan/tests/conftest.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def build_pymongo_backed_broker_with_imgs(request):
     mds_test_conf = dict(database=db_name, host='localhost',
                          port=27017, timezone='US/Eastern')
     exp_mds = MDS(mds_test_conf, 1,
-                  # auth=False
+                  auth=False
```
Comment (CJ-Wright) [inline xpdan/tests/conftest.py]: Try, except here as well

File: xpdan/glbl.py [hunk 4]
```diff
@@ -19,11 +19,17 @@
...
-    BASE_DIR = tempfile.mkdtemp()
+env_code = os.environ.get('ENV_VAR', None)
+if env_code:
+    # test
+    if int(env_code) == 1:
+        BASE_DIR = tempfile.mkdtemp()
+    # simulation
+    elif int(env_code) == 2:
+        BASE_DIR = os.getcwd()
+# FIXME: make ENV_VAR exist universally
```
Comment (CJ-Wright) [inline xpdan/glbl.py]: Can we compress this to `if, elif, else`?

File: xpdan/data_reduction.py [hunk 5]
```diff
@@ -268,11 +273,11 @@ def integrate_last(polarization_factor=0.99, root_dir=None,
             instance of class that handles data process, don't change it
             unless needed.
     """
-    integrate(db[-1],
-                    polarization_factor=polarization_factor,
-                    root_dir=root_dir,
-                    config_dict=config_dict,
-                    handler=handler)
+    integrate(an_glbl.db[-1],
```
Comment (CJ-Wright) [inline xpdan/data_reduction.py]: Can we pull this from the handler?

File: xpdan/data_reduction.py [hunk 6]
```diff
@@ -373,4 +378,4 @@ def save_last_tiff(dark_sub=True, max_count=None, dryrun=False):
         if set to True, file won't be saved. default is False
     """
 
-    save_tiff(db[-1], dark_sub, max_count, dryrun)
+    save_tiff(an_glbl.db[-1], dark_sub, max_count, dryrun)
```
Comment (CJ-Wright) [inline xpdan/data_reduction.py]: Shouldn't this also take a handler?

File: xpdan/999-load.py [hunk 7]
```diff
@@ -1,6 +1,91 @@
...
+from bluesky.examples import Reader, motor
+
+# set env_var
+# expected internal code: 
+# 0 -> beamline
+# 1 -> test
+# 2 -> simulation
+os.environ['ENV_VAR'] = str(2)
+
+from xpdan.glbl import an_glbl
```
Comment (CJ-Wright) [inline xpdan/999-load.py]: We might make the global factory take in an arg which specifies the type of setup to do. That way we can specify the type of setup even without an env_var.

Comment (chiahaoliu) [inline xpdan/999-load.py]: probably won't work nicely under current structure, as all directories are decided before goes into ``Glbl`` attributes. 

I can revisit it for next release.

File: run_tests.py [hunk 8]
```diff
@@ -14,7 +14,7 @@
     # args.extend(['--cov', 'xpdAcq'])
     # call pytest and exit with the return code from pytest so that
     # travis will fail correctly if tests fail
-    os.environ['IS_TEST'] = str(1)
+    os.environ['ENV_VAR'] = str(1)
```
Comment (CJ-Wright) [inline run_tests.py]: You might want to go with something more specific, this environmental variable could have been taken

File: xpdan/tests/conftest.py [hunk 9]
```diff
@@ -55,9 +55,11 @@ def build_pymongo_backed_broker(request):
     db_name = "mds_testing_disposable_{}".format(str(uuid.uuid4()))
     mds_test_conf = dict(database=db_name, host='localhost',
                          port=27017, timezone='US/Eastern')
-    mds = MDS(mds_test_conf, 1,
-              # auth=False
-              )
+    try:
+       # nasty details: to save MacOS user
+        mds = MDS(mds_test_conf, 1, auth=False)
+    except:
```
Comment (CJ-Wright) [inline xpdan/tests/conftest.py]: I would put a `TypeError` or some other exception here, just so we don't have a super broad exception.

File: run_tests.py (hunk 8 above)
Comment (CJ-Wright) [inline run_tests.py]: Would `XPDAN_SETUP` work?

File: xpdan/tests/conftest.py (hunk 9 above)
Comment (chiahaoliu) [inline xpdan/tests/conftest.py]: I put a ``KeyError``, which was the direct error from ``metadatastore``

Comment (chiahaoliu) [inline xpdan/tests/conftest.py]: and it turns out to be ``TypeError`` at the first layer

**Pull request merged**